backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java

```diff
@@ -280,12 +300,101 @@ public String getUrlForDatasetVersion(String datasetVersionId, String method)
     }
   }
 
+  /**
+   * For getting datasetVersions that user has access to (either as owner or a collaborator), fetch
+   * all datasetVersions of the requested datasetVersionIds then iterate that list and check if
+   * datasetVersion is accessible or not. The list of accessible datasetVersionIDs is built and
+   * returned by this method.
+   *
+   * @param requestedDatasetVersionIds : datasetVersion Ids
+   * @return List<String> : list of accessible DatasetVersion Id
+   * @throws InvalidProtocolBufferException InvalidProtocolBufferException
+   */
+  public List<String> getAccessibleDatasetVersionIDs(
+      List<String> requestedDatasetVersionIds, ModelDBServiceActions modelDBServiceActions)
+      throws InvalidProtocolBufferException {
+    List<DatasetVersion> datasetVersionList =
+        getDatasetVersionsByBatchIds(requestedDatasetVersionIds);
+    if (datasetVersionList.size() == 0) {
+      Status statusMessage =
+          Status.newBuilder()
+              .setCode(Code.PERMISSION_DENIED_VALUE)
+              .setMessage(
+                  ModelDBMessages.ACCESS_IS_DENIDE_DATASET_VERSION_ENTITIES_MSG
+                      + requestedDatasetVersionIds)
+              .build();
+      throw StatusProto.toStatusRuntimeException(statusMessage);
+    }
+    Map<String, String> datasetIdDatasetVersionIdMap = new HashMap<>();
+    for (DatasetVersion datasetVersion : datasetVersionList) {
+      datasetIdDatasetVersionIdMap.put(datasetVersion.getId(), datasetVersion.getDatasetId());
+    }
+    Set<String> datasetIdSet = new HashSet<>(datasetIdDatasetVersionIdMap.values());
+
+    List<String> accessibleDatasetVersionIds = new ArrayList<>();
+    List<String> allowedDatasetIds;
+    // Validate if current user has access to the entity or not
+    if (datasetIdSet.size() == 1) {
+      roleService.isSelfAllowed(
```

Comment:
We should check if the user has permission for the dataset version too. I don't recall the specifics of permission, but I recall that for experiment run we have a role binding just to allow the owner of an experiment run to delete it while being collaborator of another project. I can imagine a similar scenario here. So we should check permission both at the dataset and dataset version level.

Reply:
@conradoverta I remember that logic but it was added in Hydrated service. actually we are adding an action list while preparing hydrated experiment run and at that time we are adding delete action in the list where a user is an owner of those ExperimentRuns for the UI to unable Delete Button but at MDB while deleting we do not check the delete permission for ExperimentRun. we used UPDATE there https://github.com/VertaAI/modeldb/blob/aj/find-dataset-version-by-workspace/backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java#L382
And that same scenario currently implemented for Dataset & dataset version also.

Reply:
>at MDB while deleting we do not check the delete permission for ExperimentRun

This is a bug. Every request must check for the appropriate permissions. @ravishetye please scan MDB and report all APIs that currently don't satisfy this.

Reply:
#447 should address this